### DISCUSSION TITLE
Problems in removing all images in PDF

### DISCUSSION THREAD

**Discussion by lunamaze** [pymupdf/PyMuPDF#3231] [Looking for help] [answered]
My objective is to remove all images in a PDF, yet doc.save() returns the original file (with size a little bit larger) while doc.save_snapshot() works.
Moreover, some of PDFs raise 'RuntimeError: code=2: Can't do incremental writes on a repaired file' when i run doc.save_snapshot(). Could anyone please give me some hints?

Version: 1.23.26
Code:
```
doc = fitz.open("Report.pdf")
for i in range(doc.page_count):
    page = doc[i]
    d = list(set(doc.get_page_images(i)))
    for item in d:
        r = page.get_image_bbox(item[7])
        page.add_redact_annot(fitz.Rect(r.x0, r.y0, r.x1, r.y1), " ", align=fitz.TEXT_ALIGN_CENTER, fill=(1, 1, 1))
    page.apply_redactions(images=1)

#doc.save("Report_new.pdf") # doesn't work
doc.save_snapshot("Report_new.pdf") # works
```

**Comment by JorjMcKie** [comment]
Several comments:
* Make sure to use `fitz.PDF_REDACT_IMAGE_REMOVE` (=1 which you did) to address any coordinate imprecisions.
* Make sure to **_physically remove the deleted images_** by using a less careless version of saving the document: either do `doc.ez_save("filename")` or, equivalently, `doc.save(filename, garbage=3, deflate=True)`. After this, no trace of the images is left over and the file size will decrease accordingly. Redaction processing under the hood produces a number of updated versions of PDF objects. If you do not save properly all these object ghosts are still there afterwards.
* Doing any form of incremental save will only change a few object pointer values, the specification of which is appended to the file (which consequently has an increased file size). No physical removal will take place at all. In addition - you commented on this - incremental save is not always possible at all.

**Reply by JorjMcKie** [reply, marked as answer]
Just for everybody's information: This question has been answered in another channel.
It was ultimately determined that using `page.delete_image(xref)` is the most adequate and efficient solution - as opposed to redaction annotations, which require much more involvement to get it right.
